pl_examples/hydra_examples/pl_template.py

```diff
@@ -0,0 +1,69 @@
+"""
+Runs a model on the CPU on a single node.
+"""
+
+import hydra
+import pl_examples.hydra_examples.conf
+from omegaconf import DictConfig
+import pytorch_lightning.trainer.trainer_conf
+from pl_examples.models.hydra_config_model import LightningTemplateModel
+from pytorch_lightning import Callback, seed_everything, Trainer
+
+
+class MyPrintingCallback(Callback):
+    def on_init_start(self, trainer):
+        print("Starting to init trainer!")
+
+    def on_init_end(self, trainer):
+        print("trainer is init now")
+
+    def on_train_end(self, trainer, pl_module):
+        print("do something when training ends")
+
+
+class MessageCallback(Callback):
+    def __init__(self, iter_num):
+        self.iter_num = iter_num
+
+    def on_batch_start(self, trainer, pl_module):
+        if trainer.batch_idx == self.iter_num:
+            print(f"Iteration {self.iter_num}")
+
+
+def conf_init(cfg, key):
+    return hydra.utils.instantiate(getattr(cfg, key)) if hasattr(cfg, key) else None
+
+
+@hydra.main(config_path="conf", config_name="config")
```

Comment:
The idea is to create a top level config and use it here.
PytorchLightningTopConfig (maybe rename to PLConfig) is a part of the core.
It makes sense for the user config to extend it.
Let's say:
```python
@dataclass
class ExampleConfig(PLConfig):
  # with some optional additional app specific fields
  extra_fields : int = 5
```

Register this one under the name config, and it will become the schema for config.yaml (at this point you can probably just drop config.yaml, your call).
It should probably inherit from PytorchLightningTopConfig (Maybe rename it to PytorchLightningConfig

Reply:
A few questions on this, so I have created the PLConfig that is extended by the Users Config. For the PL Config, I created a default for trainer, but the others I would like to be null unless set, is the format in the most recent commit correct for this? Another thing, when I use the User Config to verify the config.yaml, I get the error
 ```
ValueError: HydraConfig was not set
	full_key: data.ds.params.runtime.cwd
	reference_type=Optional[Dict[Any, Any]]
	object_type=dict

	full_key:
	reference_type=Optional[UserConfig]
	object_type=UserConfig
```
which is from the env variable not being substituted: ${hydra:runtime.cwd}. Is this a bug, or is there a better way to structure this config?

Reply:
1. The format is okay but it's a bit of an undocumented feature. you can just omit them I think.
The only advantage of specifying them with null is that when you are overriding them from the command line it would be an override and not an addition to the defaults list so you will not need to prefix with a +.

2. Could be a bug, can you create a minimal repro and file an issue?

Reply:
Sorry I was OOL for like 2 days, trying to get something ready for upcoming deadlines.

Is the philosophy of this change to include the conf directly in the top level script as a dataclass, rather than as a file in `conf`?
Also the changes, broke my ability to run `pl_example.py` as there is no longer a `config.yaml` and it is looking for a `pl` key:
```
Key 'pl' not in 'UserConfig'
        full_key: pl
        reference_type=Optional[UserConfig]
        object_type=UserConfig
```

Reply:
also, in the vein of clarity at the top level, do you think it would improve readability to move `def conf_init` into a different file since it really just seems like a util that the user doesn't need to interact with?

Reply:
in general, after commit `a5679`, I feel that the toplevel file becomes way less readable and therefore less likely to be portable or extensible (if those are things we care about) - arguably one of the important features PL enables

Reply:
@romesc-CMU:

The philosophy is to provide type safety to the entire config.
Hydra supports both modes:
1. Config is purely in a dataclass
2. Config schema and some defaults are in a dataclass and that dataclass is validating a config file.

> Also the recent changes broke my ability to run pl_example.py as there is no longer a config.yaml and it is looking for a pl key

A bug, @anthonytec2?

> also, in the vein of clarity at the top level, do you think it would improve readability to move def conf_init into a different file since it really just seems like a util that the user doesn't need to interact with?

This will no longer be required soon. instantiate(None) would return None and you will be able to avoid having it alltogether.

> in general, after commit a5679, I feel that pl_template.py becomes a bit less readable (for a novice PL user) and therefore less likely to be portable or extensible (if those are things we care about) - arguably one of the important features PL enables

We should address this with documentation and examples.
It is not less flexible than the previous version, and it does offer better developer experience because there is top level type safety.

We can also mix the two modes. feel free to play with it and suggest an alternative.

Reply:
Yes, the PL issue is a bug I recently fixed. Although, I still have a bug due to the data location and string interpolation. Please hard code root to "." in the data group for now to run .